# Домашнє завдання до теми "Розширені методи обробки даних у Pandas"

В цьому домашньому завданні продовжуємо працювати з набором даних `supermarket_sales.csv`.

0. Імпортуйте бібліотеку pandas та зчитайте дані у змінну `df` типу `pandas.DataFrame`.

In [1]:
import pandas as pd

In [2]:
df = pd.read_csv('../Data/supermarket_sales.csv')

1. Відсортуйте дані за спаданням ціни товара ('Unit price') і виведіть 5 перших рядочків на екран.

In [5]:
df.sort_values(by = 'Unit price', ascending=False).head(5)

Invoice ID Branch       City Customer type Gender         Product line  \
983  148-41-7930      C  Naypyitaw        Normal   Male    Health and beauty   
122  219-22-9386      B   Mandalay        Member   Male    Sports and travel   
930  641-62-7288      B   Mandalay        Normal   Male   Home and lifestyle   
494  437-53-3084      B   Mandalay        Normal   Male  Fashion accessories   
283  667-92-0055      A     Yangon        Member   Male    Health and beauty   

     Unit price  Quantity  Tax 5%    Total       Date   Time      Payment  \
983       99.96         7  34.986  734.706  1/23/2019  10:33         Cash   
122       99.96         9  44.982  944.622   3/9/2019  17:26  Credit card   
930       99.92         6  29.976  629.496  3/24/2019  13:33      Ewallet   
494       99.89         2   9.989  209.769  2/26/2019  11:48      Ewallet   
283       99.83         6  29.949  628.929   3/4/2019  15:02      Ewallet   

       cogs  gross margin percentage  gross income  Rating  
983  699.72                 4.761905        34.986     6.1  
122  899.64                 4.761905        44.982     4.2  
930  599.52                 4.761905        29.976     7.1  
494  199.78                 4.761905         9.989     7.1  
283  598.98                 4.761905        29.949     8.5

2. Знайдіть інвойс (рядок в даних) в категорії товарів Electronic accessories, в якому був куплений товар з найбільшою кількістю і найбільшою ціною. Виведіть цей рядок даних на екран.

In [14]:
df[df['Product line'] == 'Electronic accessories'].sort_values(by = 'Unit price', ascending=False).head(1)

Invoice ID Branch      City Customer type  Gender  \
209  817-69-8206      B  Mandalay        Normal  Female   

               Product line  Unit price  Quantity   Tax 5%     Total  \
209  Electronic accessories       99.73         9  44.8785  942.4485   

         Date   Time      Payment    cogs  gross margin percentage  \
209  3/2/2019  19:42  Credit card  897.57                 4.761905   

     gross income  Rating  
209       44.8785     6.5

In [15]:
df[df['Product line'] == 'Electronic accessories'].sort_values(by = 'Quantity', ascending=False).head(1)

Invoice ID Branch    City Customer type Gender            Product line  \
105  704-48-3927      A  Yangon        Member   Male  Electronic accessories   

     Unit price  Quantity  Tax 5%    Total       Date   Time  Payment   cogs  \
105       88.67        10  44.335  931.035  1/12/2019  14:50  Ewallet  886.7   

     gross margin percentage  gross income  Rating  
105                 4.761905        44.335     7.3

Отже, товар з Invoice ID 817-69-8206 був куплений за найбільшою ціною: 99.73; товар з Invoice ID 704-48-3927 був куплений у найбільшій кількості: 10

3.1. Створіть нову колонку в даних `Rating Category`, яка визначає категорію рейтингу: високий, середній чи низький за наступним правилом:
- якщо рейтинг вищий або рівний 8 - то він високий (`High`)
- якщо рейтинг від 6 до 8 не включаючи - то він середній (`Medium`)
- інакше - низький (`Low`).

In [24]:
import random
values = [random.choice([1, 2, 3, 4, 5, 6, 7, 8]) for _ in range(len(df))]
df['Rating'] = values

In [28]:
def categories(rating_value):
    if rating_value >= 8:
        return 'High'
    elif 6 <= rating_value < 8:
        return 'Medium'
    else:
        return 'Low'

df['Rating category'] = df.Rating.apply(categories)

In [31]:
df[['Rating', 'Rating category']].head(5)

Rating Rating category
0       4             Low
1       6          Medium
2       5             Low
3       4             Low
4       8            High

3.2. Виведіть, який відсоток даних становить кожне зі значень категорій рейтингу.

In [34]:
df['Rating category'].value_counts(normalize=True)


Rating category
Low       0.644
Medium    0.243
High      0.113
Name: proportion, dtype: float64

4.1. Запустіть код нижче аби створити змінну `time`. Ця змінна містить час в форматі "година:хвилина". Напишіть функцію `get_hour`, яка буде зі змінної формату як `time` дістати лише години і конвертувати їх у тип `int`.

Запустіть написану функцію на змінній `time` та перевірте правильність роботи функції.

In [35]:
time = '11:28'

In [38]:
def get_hour(time_hours):
    hour = int(time_hours.split(':')[0])
    return hour

hour = get_hour(time)
hour

11

4.2. З допомогою функції `get_hour` і метода `DataFrame.apply()` створіть нову колонку `Hour` в нашому наборі даних, яка буде містити години здійснення покупки, обчислені на основі часу здійснення покупки `Time`.

In [40]:
df['Hour'] = df.Time.apply(get_hour)
df['Hour'].

0      13
1      10
2      13
3      20
4      10
       ..
995    13
996    17
997    13
998    15
999    13
Name: Hour, Length: 1000, dtype: int64

4.3. Виведіть кількість значень за кожною годиною покупки і дайте відповідь "О яких годинах найбільше здійснюють покупок покупці і чому на Ваш погляд саме о цих годинах? Чи це логічно, що найбільше покупок саме в ці години?".

In [61]:
df[['Quantity', 'Hour']]

Quantity  Hour
0           7    13
1           5    10
2           7    13
3           8    20
4           7    10
..        ...   ...
995         1    13
996        10    17
997         1    13
998         1    15
999         7    13

[1000 rows x 2 columns]

In [62]:
df[['Quantity', 'Hour']].max()

Quantity    10
Hour        20
dtype: int64

Виходить, що найбільше покупок здійснюється ввечері. Як на мене, це через те, що люди після роботи заходять у супермаркет. Тому так, на мій погляд, усе логічно

5. Виведіть середнє значення чеку (`Total`) для покупок в містах, наявних в даних, за допомогою функції `groupby`.

In [66]:
df.Total.groupby(by = df['City']).mean()

City
Mandalay     319.872506
Naypyitaw    337.099715
Yangon       312.354031
Name: Total, dtype: float64

6. Виведіть кількість покупок, середнє, мінімальне та максимальне значення для загальної суми чеку (Total) в містах, наявних в даних, за допомогою функції groupby.
Зробіть висновки з цих показників та дайте рекомендації для мережі супермаркетів.

In [76]:
df.Total.groupby(by = df['City']).agg(['sum', 'mean', 'min', 'max'])

sum        mean      min      max
City                                                
Mandalay   106197.6720  319.872506  18.6375  1022.49
Naypyitaw  110568.7065  337.099715  10.6785  1042.65
Yangon     106200.3705  312.354031  12.6945  1039.29

Судячі з отриманих результатів, найбільше попупок здійснюється у місті Naypyitaw. Для того, щоб зробити висновки, на мій погляд, варто дослідити, які саме покупки здійснюють покупці, можливо супермаркети у цьомі місті проводят якусь спеціальну політику заохочення клієнтів, а також категорії покупців, що здійснюють з найбільшою загальною сумою чеку.

7. Ми внесли деякі зміни в наш набір даних, тож, давайте запишемо результат у файл. Запишіть оновлену версію даних у файл `supermarket_sales_enriched.csv`. Зверніть увагу, аби не записати нічого зайвого, наприклад, індекси рядків нам в файлі не потрібні, оскільки вони стандартні.

In [77]:
df.to_csv('supermarket_sales_enriched.csv', index=False)

In [78]:
pd.read_csv('supermarket_sales_enriched.csv', nrows = 5)

Invoice ID Branch       City Customer type  Gender  \
0  750-67-8428      A     Yangon        Member  Female   
1  226-31-3081      C  Naypyitaw        Normal  Female   
2  631-41-3108      A     Yangon        Normal    Male   
3  123-19-1176      A     Yangon        Member    Male   
4  373-73-7910      A     Yangon        Normal    Male   

             Product line  Unit price  Quantity   Tax 5%     Total       Date  \
0       Health and beauty       74.69         7  26.1415  548.9715   1/5/2019   
1  Electronic accessories       15.28         5   3.8200   80.2200   3/8/2019   
2      Home and lifestyle       46.33         7  16.2155  340.5255   3/3/2019   
3       Health and beauty       58.22         8  23.2880  489.0480  1/27/2019   
4       Sports and travel       86.31         7  30.2085  634.3785   2/8/2019   

    Time      Payment    cogs  gross margin percentage  gross income  Rating  \
0  13:08      Ewallet  522.83                 4.761905       26.1415       4   
1  10:29         Cash   76.40                 4.761905        3.8200       6   
2  13:23  Credit card  324.31                 4.761905       16.2155       5   
3  20:33      Ewallet  465.76                 4.761905       23.2880       4   
4  10:37      Ewallet  604.17                 4.761905       30.2085       8   

  Rating category  Hour  
0             Low    13  
1          Medium    10  
2             Low    13  
3             Low    20  
4            High    10